File: examples/typescript/read_snapshot.ts
import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import { fileURLToPath } from "node:url";

type JsonObject = Record<string, unknown>;
export type Direction = "bullish" | "bearish" | "neutral";
export interface SnapshotRecord extends JsonObject {
  id: string;
  concept: string;
  direction: Direction;
  state: string;
  lower: number;
  upper: number;
}
export interface Snapshot extends JsonObject {
  status: string;
  symbol: string;
  timeframe: string;
  as_of: string | null;
  records: SnapshotRecord[];
}
export interface Filters {
  concept?: string;
  direction?: Direction;
}

function invalid(path: string, expectation: string): never {
  throw new Error(`${path}: expected ${expectation}`);
}
function object(value: unknown, path: string): JsonObject {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    return invalid(path, "an object");
  }
  return value as JsonObject;
}
function nonempty(value: unknown, path: string): string {
  if (typeof value !== "string" || value.length === 0) {
    return invalid(path, "a nonempty string");
  }
  return value;
}
function enumeration(value: unknown, path: string, values: readonly string[]): string {
  if (typeof value !== "string" || !values.includes(value)) {
    return invalid(path, values.join(" | "));
  }
  return value;
}
function finite(value: unknown, path: string): number {
  if (typeof value !== "number" || !Number.isFinite(value)) {
    return invalid(path, "a finite number");
  }
  return value;
}
function boolean(value: unknown, path: string): void {
  if (typeof value !== "boolean") invalid(path, "a boolean");
}
function string(value: unknown, path: string): void {
  if (typeof value !== "string") invalid(path, "a string");
}
function integer(value: unknown, path: string, minimum: number, maximum = Infinity): void {
  const number = finite(value, path);
  if (!Number.isInteger(number) || number < minimum || number > maximum) {
    invalid(path, `an integer between ${minimum} and ${maximum}`);
  }
}
function array(value: unknown, path: string): unknown[] {
  if (!Array.isArray(value)) return invalid(path, "an array");
  return value;
}

/** Broker wall-clock values carry no UTC suffix or inferred timezone. */
function timestamp(value: unknown, path: string, nullable = false): void {
  if (nullable && value === null) return;
  if (typeof value !== "string" || !/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}$/.test(value)) {
    invalid(path, `a broker timestamp YYYY-MM-DDTHH:mm:ss${nullable ? " or null" : ""}`);
  }
  const [year, month, day, hour, minute, second] = (value as string).split(/[-T:]/).map(Number);
  const leapYear = year! % 4 === 0 && (year! % 100 !== 0 || year! % 400 === 0);
  const days = [31, leapYear ? 29 : 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31][month! - 1] ?? 0;
  if (year! < 1 || month! < 1 || month! > 12 || day! < 1 || day! > days ||
      hour! > 23 || minute! > 59 || second! > 59) {
    invalid(path, "a valid broker calendar date and time");
  }
}

const statuses = ["READY", "PARTIAL", "NOT_READY", "ERROR", "DISABLED"] as const;
const directions = ["bullish", "bearish", "neutral"] as const;
const concepts = [
  "SWING_HIGH", "SWING_LOW", "BOS", "CHOCH", "ORDER_BLOCK", "FVG", "LIQUIDITY",
  "PREMIUM_DISCOUNT", "OTE", "KILL_ZONE", "BREAKER", "DISPLACEMENT", "MSS", "IFVG",
  "BPR", "PREVIOUS_DAY_HIGH", "PREVIOUS_DAY_LOW", "PREVIOUS_WEEK_HIGH", "PREVIOUS_WEEK_LOW",
  "SESSION_HIGH", "SESSION_LOW", "DAILY_GAP", "WEEKLY_GAP", "SMT", "PO3",
] as const;

function validateConfig(input: unknown): void {
  const config = object(input, "config");
  const limits: Record<string, number> = {
    lookback_bars: 100000, max_records_per_concept: 100000, swing_strength: 5000,
    displacement_baseline: 5000, max_zone_age: 10000, bpr_max_separation: 10000, po3_expiry_bars: 10000,
  };
  for (const [field, maximum] of Object.entries(limits)) {
    integer(config[field], `config.${field}`, 1, maximum);
  }
  integer(config.smt_radius, "config.smt_radius", 0, 5000);
  for (const field of ["displacement_multiplier", "displacement_body_fraction"]) {
    if (finite(config[field], `config.${field}`) <= 0) invalid(`config.${field}`, "a number greater than 0");
  }
  if ((config.displacement_body_fraction as number) > 1) invalid("config.displacement_body_fraction", "a number <= 1");
  if (finite(config.min_fvg_pips, "config.min_fvg_pips") < 0) invalid("config.min_fvg_pips", "a number >= 0");
  string(config.smt_symbol, "config.smt_symbol");
  for (const field of ["enable_draw", "enable_calendar", "enable_smt", "enable_po3",
    "enable_displacement", "enable_mss", "enable_ifvg", "enable_bpr", "enable_cs", "enable_vix"]) {
    boolean(config[field], `config.${field}`);
  }
  if (config.enable_smt === true) nonempty(config.smt_symbol, "config.smt_symbol");
  const sessions = array(config.sessions, "config.sessions");
  if (sessions.length !== 3) invalid("config.sessions", "exactly three sessions");
  sessions.forEach((entry, index) => {
    const path = `config.sessions[${index}]`;
    const session = object(entry, path);
    nonempty(session.name, `${path}.name`);
    integer(session.start_minute, `${path}.start_minute`, 0, 1439);
    integer(session.end_minute, `${path}.end_minute`, 0, 1439);
  });
}

export function validateSnapshot(input: unknown): Snapshot {
  const value = object(input, "snapshot");
  if (typeof value.schema_version !== "string" || !/^1\.\d+$/.test(value.schema_version)) {
    invalid("schema_version", "a supported 1.x version string");
  }
  for (const field of ["library_version", "symbol", "timeframe"]) nonempty(value[field], field);
  enumeration(value.time_basis, "time_basis", ["broker"]);
  enumeration(value.status, "status", statuses);
  if (Object.hasOwn(value, "message")) string(value.message, "message");
  timestamp(value.as_of, "as_of", true);
  validateConfig(value.config);
  array(value.modules, "modules").forEach((entry, index) => {
    const path = `modules[${index}]`;
    const module = object(entry, path);
    enumeration(module.concept, `${path}.concept`, concepts);
    enumeration(module.status, `${path}.status`, statuses);
    timestamp(module.as_of, `${path}.as_of`, true);
    boolean(module.truncated, `${path}.truncated`);
    string(module.message, `${path}.message`);
  });
  const ids = new Set<string>();
  array(value.records, "records").forEach((entry, index) => {
    const path = `records[${index}]`;
    const record = object(entry, path);
    for (const field of ["id", "state"]) nonempty(record[field], `${path}.${field}`);
    const id = record.id as string;
    if (ids.has(id)) invalid(`${path}.id`, "a unique record id");
    ids.add(id);
    enumeration(record.concept, `${path}.concept`, concepts);
    for (const field of ["source_time", "confirmed_at", "updated_at"]) timestamp(record[field], `${path}.${field}`);
    for (const field of ["period_start", "period_end"]) timestamp(record[field], `${path}.${field}`, true);
    enumeration(record.direction, `${path}.direction`, directions);
    const lower = finite(record.lower, `${path}.lower`);
    const upper = finite(record.upper, `${path}.upper`);
    if (lower > upper) invalid(path, "lower <= upper");
    boolean(record.active, `${path}.active`);
    array(record.related_ids, `${path}.related_ids`).forEach((id, item) => nonempty(id, `${path}.related_ids[${item}]`));
    for (const field of ["reference_price", "comparison_price", "strength"]) {
      finite(record[field], `${path}.${field}`);
    }
    string(record.reason, `${path}.reason`);
  });
  return value as Snapshot;
}

export function renderSnapshot(snapshot: Snapshot, filters: Filters = {}): string {
  const lines = [
    `status=${snapshot.status} symbol=${snapshot.symbol} timeframe=${snapshot.timeframe} as_of=${snapshot.as_of ?? "null"} time_basis=broker`,
  ];
  const records = snapshot.records.filter((record) =>
    (filters.concept === undefined || record.concept === filters.concept) &&
    (filters.direction === undefined || record.direction === filters.direction),
  ).sort((left, right) => left.id < right.id ? -1 : left.id > right.id ? 1 : 0);
  for (const record of records) {
    lines.push([record.id, record.concept, record.direction, record.state,
      fixedPrice(record.lower), fixedPrice(record.upper)].join("\t"));
  }
  return lines.join("\n") + "\n";
}

function fixedPrice(value: number): string {
  // Round the exact binary64 value, avoiding an intermediate floating-point
  // multiplication or decimal conversion that could move a rounding boundary.
  const bytes = new DataView(new ArrayBuffer(8));
  bytes.setFloat64(0, value, false);
  const bits = bytes.getBigUint64(0, false);
  const negative = (bits >> 63n) !== 0n;
  const encodedExponent = Number((bits >> 52n) & 0x7ffn);
  if (encodedExponent === 0x7ff) invalid("price", "a finite number");
  const fraction = bits & ((1n << 52n) - 1n);
  const significand = encodedExponent === 0 ? fraction : fraction | (1n << 52n);
  const exponent = encodedExponent === 0 ? -1074 : encodedExponent - 1023 - 52;
  const scale = 100000000n;
  const numerator = significand * scale;
  let rounded: bigint;
  if (exponent >= 0) {
    rounded = numerator << BigInt(exponent);
  } else {
    const denominator = 1n << BigInt(-exponent);
    rounded = numerator / denominator;
    const twiceRemainder = (numerator % denominator) * 2n;
    if (twiceRemainder > denominator ||
        (twiceRemainder === denominator && (rounded & 1n) !== 0n)) {
      rounded += 1n;
    }
  }
  const sign = negative && rounded !== 0n ? "-" : "";
  return `${sign}${rounded / scale}.${(rounded % scale).toString().padStart(8, "0")}`;
}

const usage = "Usage: node dist/read_snapshot.js SNAPSHOT.json [--concept CONCEPT] [--direction bullish|bearish|neutral]";
export function parseArguments(args: string[]): { path: string; filters: Filters } {
  const filters: Filters = {};
  let path: string | undefined;
  for (let index = 0; index < args.length; index++) {
    const argument = args[index]!;
    if (argument === "--concept" || argument === "--direction") {
      const value = args[++index];
      if (value === undefined || value.startsWith("--")) throw new Error(`${argument} requires a value`);
      if (argument === "--concept") {
        if (filters.concept !== undefined) throw new Error("--concept may only be specified once");
        if (!/^[A-Z][A-Z0-9_]*$/.test(value)) throw new Error("--concept must be an uppercase concept name");
        filters.concept = value;
      } else {
        if (filters.direction !== undefined) throw new Error("--direction may only be specified once");
        filters.direction = enumeration(value, "--direction", directions) as Direction;
      }
    } else if (argument.startsWith("--")) {
      throw new Error(`Unknown option: ${argument}`);
    } else if (path === undefined) {
      path = argument;
    } else {
      throw new Error("Only one snapshot path is allowed");
    }
  }
  if (path === undefined) throw new Error(usage);
  return { path, filters };
}

function main(): void {
  if (process.argv.slice(2).includes("--help")) {
    process.stdout.write(usage + "\n");
    return;
  }
  try {
    const { path, filters } = parseArguments(process.argv.slice(2));
    const snapshot = validateSnapshot(JSON.parse(readFileSync(path, "utf8")) as unknown);
    process.stdout.write(renderSnapshot(snapshot, filters));
  } catch (error) {
    process.stderr.write(`Error: ${error instanceof Error ? error.message : String(error)}\n`);
    process.exitCode = 1;
  }
}

if (process.argv[1] !== undefined && resolve(process.argv[1]) === fileURLToPath(import.meta.url)) {
  main();
}

File: examples/python/read_snapshot.py
#!/usr/bin/env python3
"""Validate and print an SMC snapshot using only the Python standard library."""

from __future__ import annotations

import argparse
import json
import math
import re
import sys
from datetime import datetime
from pathlib import Path
from typing import Any

STATUSES = {"READY", "PARTIAL", "NOT_READY", "ERROR", "DISABLED"}
DIRECTIONS = {"bullish", "bearish", "neutral"}
TIMESTAMP = re.compile(r"\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\Z")
CONCEPT = re.compile(r"[A-Z][A-Z0-9_]*\Z")
CONCEPTS = {
    "SWING_HIGH", "SWING_LOW", "BOS", "CHOCH", "ORDER_BLOCK", "FVG", "LIQUIDITY",
    "PREMIUM_DISCOUNT", "OTE", "KILL_ZONE", "BREAKER", "DISPLACEMENT", "MSS", "IFVG",
    "BPR", "PREVIOUS_DAY_HIGH", "PREVIOUS_DAY_LOW", "PREVIOUS_WEEK_HIGH",
    "PREVIOUS_WEEK_LOW", "SESSION_HIGH", "SESSION_LOW", "DAILY_GAP", "WEEKLY_GAP",
    "SMT", "PO3",
}
POSITIVE_INTEGERS = (
    "lookback_bars", "max_records_per_concept", "swing_strength", "displacement_baseline",
    "max_zone_age", "bpr_max_separation", "po3_expiry_bars",
)
INTEGER_MAXIMA = (100000, 100000, 5000, 5000, 10000, 10000, 10000)
FLAGS = (
    "enable_draw", "enable_calendar", "enable_smt", "enable_po3", "enable_displacement",
    "enable_mss", "enable_ifvg", "enable_bpr", "enable_cs", "enable_vix",
)


class SnapshotError(ValueError):
    """The file does not implement the supported snapshot contract."""


def require(condition: bool, message: str) -> None:
    if not condition:
        raise SnapshotError(message)


def object_value(value: Any, path: str) -> dict[str, Any]:
    require(isinstance(value, dict), f"{path} must be an object")
    return value


def fields(value: Any, names: tuple[str, ...], path: str) -> dict[str, Any]:
    obj = object_value(value, path)
    for name in names:
        require(name in obj, f"{path}.{name} is required")
    return obj


def string(value: Any, path: str, *, nonempty: bool = True) -> None:
    require(isinstance(value, str), f"{path} must be a string")
    require(not nonempty or len(value) > 0, f"{path} must not be empty")


def number(value: Any, path: str, *, nullable: bool = False) -> None:
    if nullable and value is None:
        return
    require(type(value) in (int, float), f"{path} must be a number")
    try:
        finite = math.isfinite(value)
    except OverflowError:
        finite = False
    require(finite, f"{path} must be finite")


def timestamp(value: Any, path: str, *, nullable: bool = False) -> None:
    if nullable and value is None:
        return
    string(value, path)
    require(TIMESTAMP.fullmatch(value) is not None,
            f"{path} must be a broker timestamp YYYY-MM-DDTHH:mm:ss without a timezone")
    try:
        datetime.fromisoformat(value)
    except ValueError as exc:
        raise SnapshotError(f"{path} is not a valid date/time") from exc


def enum(value: Any, allowed: set[str], path: str) -> None:
    string(value, path)
    require(value in allowed, f"{path} must be one of {', '.join(sorted(allowed))}")


def integer(value: Any, path: str, minimum: int, maximum: int | None = None) -> None:
    number(value, path)
    require(value == int(value), f"{path} must be an integer")
    require(value >= minimum, f"{path} must be at least {minimum}")
    require(maximum is None or value <= maximum, f"{path} must not exceed {maximum}")


def validate_config(config: Any) -> None:
    fields(config, POSITIVE_INTEGERS + FLAGS + ("smt_radius", "smt_symbol", "sessions",
           "displacement_multiplier", "displacement_body_fraction", "min_fvg_pips"), "config")
    for name, maximum in zip(POSITIVE_INTEGERS, INTEGER_MAXIMA, strict=True):
        integer(config[name], f"config.{name}", 1, maximum)
    integer(config["smt_radius"], "config.smt_radius", 0, 5000)
    for name in FLAGS:
        require(type(config[name]) is bool, f"config.{name} must be boolean")
    for name in ("displacement_multiplier", "displacement_body_fraction", "min_fvg_pips"):
        number(config[name], f"config.{name}")
    require(config["displacement_multiplier"] > 0, "config.displacement_multiplier must be positive")
    require(0 < config["displacement_body_fraction"] <= 1,
            "config.displacement_body_fraction must be in (0, 1]")
    require(config["min_fvg_pips"] >= 0, "config.min_fvg_pips must be nonnegative")
    string(config["smt_symbol"], "config.smt_symbol", nonempty=False)
    require(not config["enable_smt"] or bool(config["smt_symbol"]),
            "config.smt_symbol must not be empty when SMT is enabled")
    require(isinstance(config["sessions"], list) and len(config["sessions"]) == 3,
            "config.sessions must be an array of three sessions")
    for index, session in enumerate(config["sessions"]):
        path = f"config.sessions[{index}]"
        fields(session, ("name", "start_minute", "end_minute"), path)
        string(session["name"], f"{path}.name")
        for name in ("start_minute", "end_minute"):
            integer(session[name], f"{path}.{name}", 0, 1439)


def reject_constant(value: str) -> None:
    raise SnapshotError(f"nonfinite JSON number {value} is not supported")


def load_snapshot(path: str | Path) -> dict[str, Any]:
    with Path(path).open(encoding="utf-8") as source:
        snapshot = json.load(source, parse_constant=reject_constant)
    validate_snapshot(snapshot)
    return snapshot


def validate_snapshot(snapshot: Any) -> None:
    fields(snapshot, ("schema_version", "library_version", "symbol", "timeframe",
                      "time_basis", "as_of", "status", "config", "modules", "records"),
           "snapshot")
    for name in ("schema_version", "library_version", "symbol", "timeframe"):
        string(snapshot[name], name)
    require(re.fullmatch(r"1\.[0-9]+", snapshot["schema_version"]) is not None,
            "schema_version must use supported major version 1 (1.<minor>)")
    require(snapshot["time_basis"] == "broker", "time_basis must be broker")
    timestamp(snapshot["as_of"], "as_of", nullable=True)
    enum(snapshot["status"], STATUSES, "status")
    if "message" in snapshot:
        string(snapshot["message"], "message", nonempty=False)
    validate_config(snapshot["config"])
    require(isinstance(snapshot["modules"], list), "modules must be an array")
    for index, module in enumerate(snapshot["modules"]):
        path = f"modules[{index}]"
        fields(module, ("concept", "status", "as_of", "truncated", "message"), path)
        enum(module["concept"], CONCEPTS, f"{path}.concept")
        enum(module["status"], STATUSES, f"{path}.status")
        timestamp(module["as_of"], f"{path}.as_of", nullable=True)
        require(type(module["truncated"]) is bool, f"{path}.truncated must be boolean")
        string(module["message"], f"{path}.message", nonempty=False)
    require(isinstance(snapshot["records"], list), "records must be an array")
    ids: set[str] = set()
    for index, record in enumerate(snapshot["records"]):
        validate_record(record, f"records[{index}]")
        require(record["id"] not in ids, f"records[{index}].id duplicates {record['id']}")
        ids.add(record["id"])


def validate_record(record: Any, path: str) -> None:
    fields(record, ("id", "concept", "source_time", "confirmed_at", "updated_at",
                    "direction", "lower", "upper", "state", "active", "related_ids",
                    "period_start", "period_end", "reference_price", "comparison_price",
                    "strength", "reason"), path)
    for name in ("id", "concept", "state"):
        string(record[name], f"{path}.{name}")
    enum(record["concept"], CONCEPTS, f"{path}.concept")
    for name in ("source_time", "confirmed_at", "updated_at"):
        timestamp(record[name], f"{path}.{name}")
    for name in ("period_start", "period_end"):
        timestamp(record[name], f"{path}.{name}", nullable=True)
    enum(record["direction"], DIRECTIONS, f"{path}.direction")
    for name in ("lower", "upper"):
        number(record[name], f"{path}.{name}")
    require(record["lower"] <= record["upper"], f"{path}.lower must not exceed upper")
    require(type(record["active"]) is bool, f"{path}.active must be boolean")
    require(isinstance(record["related_ids"], list), f"{path}.related_ids must be an array")
    for index, related_id in enumerate(record["related_ids"]):
        string(related_id, f"{path}.related_ids[{index}]")
    for name in ("reference_price", "comparison_price", "strength"):
        number(record[name], f"{path}.{name}")
    string(record["reason"], f"{path}.reason", nonempty=False)


def format_price(value: float) -> str:
    """Format the binary64 value with round-to-even; rendered zero has no sign."""
    rendered = f"{float(value):.8f}"
    return "0.00000000" if rendered == "-0.00000000" else rendered


def render_snapshot(snapshot: dict[str, Any], concept: str | None = None,
                    direction: str | None = None) -> str:
    as_of = snapshot["as_of"] if snapshot["as_of"] is not None else "null"
    lines = [(f"status={snapshot['status']} symbol={snapshot['symbol']} "
              f"timeframe={snapshot['timeframe']} as_of={as_of} time_basis=broker")]
    for record in sorted(snapshot["records"], key=lambda item: item["id"]):
        if concept is not None and record["concept"] != concept:
            continue
        if direction is not None and record["direction"] != direction:
            continue
        lines.append("\t".join((record["id"], record["concept"], record["direction"],
                                record["state"], format_price(record["lower"]),
                                format_price(record["upper"]))))
    return "\n".join(lines) + "\n"


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("snapshot", help="UTF-8 snapshot JSON path")
    parser.add_argument("--concept", help="uppercase concept, for example IFVG")
    parser.add_argument("--direction", choices=sorted(DIRECTIONS))
    args = parser.parse_args(argv)
    if args.concept is not None and CONCEPT.fullmatch(args.concept) is None:
        parser.error("--concept must be an uppercase concept name")
    try:
        snapshot = load_snapshot(args.snapshot)
        sys.stdout.write(render_snapshot(snapshot, args.concept, args.direction))
        return 0
    except (OSError, UnicodeError, ValueError) as exc:
        print(f"error: {exc}", file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())
